File: server/state.py
"""The game session and the helpers every handler needs.

Handlers live in `handlers/` and import from here. Everything that changes
during a session — the game, the lobby's rule selection, who is watching, the
history, the flood limiters — belongs to a `GameSession` reached through
`state.session()`. It used to be a set of module globals, two of which were
*rebound* rather than mutated, so every handler had to remember to spell them
`state.current_game` or quietly keep looking at a finished game. Attributes on
one object cannot go stale that way, and a test can build a fresh session
instead of unsetting globals one by one.
"""

import json
import logging
import os
import random
import threading

import build_info
from config import get_config
from extensions import socketio
from flask import request
from flask_socketio import disconnect, emit
from game import commands as commands_module
from game import persistence
from game import rules as rules_module
from game.event_log import EventLog
from game.rate_limit import (
    AbuseTracker,
    RateLimiter,
    limit_for,
    payload_too_large,
)

logger = logging.getLogger(__name__)

config = get_config()

DATA_FILE = os.path.join(config.DATA_DIR, 'users.json')
SAVE_FILE = os.path.join(config.DATA_DIR, 'game.json')
MAX_PLAYERS = config.MAX_PLAYERS
MIN_PLAYERS = config.MIN_PLAYERS

# Serializes writes to users.json. Not part of the session: the file outlives
# every game, and a new session must not hand out a second lock on it.
users_file_lock = threading.Lock()


class GameSession:
    """One server's worth of mutable state.

    There is exactly one game per process — state lives in memory and the
    deployment is a single worker (see wsgi.py) — so there is exactly one of
    these. It exists so that the game and everything that travels with it are
    attributes reached through `session()` rather than module globals that get
    rebound underneath the handlers reading them.
    """

    def __init__(self, config=None):
        self.config = config or get_config()
        self.game = None

        # Rules the lobby has selected for the next game. Anyone in the lobby
        # can change these; they are frozen into the Game when it starts,
        # because a rule that changed mid-game would invalidate decisions
        # players had already made.
        self.lobby_rules = rules_module.defaults()

        # Socket id -> the seat this connection holds. It decides both which
        # private hand the socket is shown and, through `require_actor`, whose
        # actions it may take: identity comes from the connection, never from
        # a `name` in the payload, which anyone can forge. Covering for someone
        # who stepped away still works — it means joining as them with
        # `takeover`, which moves the seat to that socket and is logged.
        self.viewers = {}

        # Chat and game history. Survives games so the table can look back at
        # what happened; bounded internally so a long session cannot grow
        # without limit.
        self.event_log = EventLog()

        # Serializes validate-then-apply for the single game. Two events
        # arriving in the same tick would otherwise both validate against the
        # pre-action state and both apply — the last development card bought
        # twice, a double-clicked road placed twice.
        self.lock = threading.RLock()

        # Flood protection. One limiter covers every event, keyed by socket and
        # event, so a client cannot burn the server down with cheap-to-send,
        # expensive-to-serve messages like refresh_board.
        self.rate_limiter = RateLimiter()
        self.abuse_tracker = AbuseTracker()


_session = GameSession()


def session() -> GameSession:
    """The session this process is serving. The one way in."""
    return _session


def new_session(config=None) -> GameSession:
    """Start again from an empty session.

    Called by the factory, so building an app means serving a fresh server —
    which is what lets a test start from a clean table instead of unsetting
    globals one at a time and hoping it remembered them all.
    """
    global _session
    _session = GameSession(config)
    return _session

def restore_saved_game():
    """Bring back an interrupted game on startup.

    A corrupt or outdated save is reported and skipped rather than half-loaded:
    resuming a game whose invariants do not hold is worse than starting fresh.
    """
    try:
        restored = persistence.load(SAVE_FILE, config=config)
    except persistence.NotASaveFile:
        logger.info("%s is not a saved game; starting fresh", SAVE_FILE)
        return
    except Exception:
        logger.exception("ignoring unusable save at %s", SAVE_FILE)
        return
    if restored is None:
        return
    session().game = restored
    logger.info("restored game: players=%s phase=%s turn=%s",
                restored.get_player_names(), restored.game_phase, restored.turn_count)

COLOR_PALETTE = [
    '#3498db',  # Blue
    '#27ae60',  # Green
    '#f1c40f',  # Yellow
    '#e74c3c',  # Red
    '#9b59b6',  # Purple
    '#e67e22',  # Orange
    '#ff6b9d',  # Pink
    '#ecf0f1',  # White
    '#2c3e50',  # Black
]


def seat_limit() -> int:
    """How many players this table seats.

    The cap is a rule like the clocks are: 0 means "whatever this server is
    configured with", so a table that never opened the setting gets the
    deployment's own number and an old save loads without inventing one. It
    was a bare config constant, which made it the one seating number the
    lobby could not agree on — `min_players` has always been a rule.

    A running game answers from its own frozen rules: the rules a game started
    under are the rules it is played under, and a seat opened mid-game by a
    lobby edit would seat a seventh player into a six-player board.
    """
    game = session().game
    rules = game.rules if game is not None else session().lobby_rules
    return rules['max_players'] or MAX_PLAYERS


def get_random_color():
    """Get a random color from the palette."""
    return random.choice(COLOR_PALETTE)

def load_users():
    if os.path.exists(DATA_FILE):
        with open(DATA_FILE) as f:
            data = json.load(f)
            return data.get('users', [])
    return []

def save_users(users):
    """Write the user list atomically.

    A plain in-place write that is interrupted, or two writes interleaving,
    leaves a truncated file that fails to parse on the next read and breaks
    joining permanently. Writing to a temp file in the same directory and
    renaming makes the replacement all-or-nothing.
    """
    os.makedirs(os.path.dirname(DATA_FILE), exist_ok=True)
    temp_path = f"{DATA_FILE}.tmp"
    with open(temp_path, 'w') as f:
        json.dump({'users': users}, f)
        f.flush()
        os.fsync(f.fileno())
    os.replace(temp_path, DATA_FILE)

def update_users(mutate):
    """Read, modify, and write the user list as one atomic unit.

    The read and the write must be under the same lock: two players joining at
    the same moment would otherwise both read the old list, each append
    themselves, and the second write would drop the first player entirely.

    `mutate` takes the current list and returns (new_list, result); result is
    passed back to the caller.
    """
    with users_file_lock:
        users, result = mutate(load_users())
        if users is not None:
            save_users(users)
        return result

def reject(code, message):
    """Send a rejection to the client whose action failed, and log it.

    Every rejected action gets a reply — a client that hears nothing cannot
    tell rejection from a dropped connection, and leaves its button spinning.
    """
    logger.warning("rejected %s from sid=%s: %s (%s)", request.event.get('message'),
                   request.sid, message, code)
    # Sender only. One player's illegal move is not the table's business, and
    # what they tried to do leaks what they hold.
    emit('error', {'code': code, 'message': message})

def rate_limited(budget_event=None) -> bool:
    """Whether this event should be dropped. Call at the top of a handler.

    Checks payload size first: deserialising hostile input is itself the denial
    of service, so an oversized message is refused before it is looked at.
    Persistent offenders are disconnected rather than merely throttled — a
    client behaving that way is not a player having a bad network day.

    `budget_event` spends another event's bucket instead of this one's. It
    exists for the slash commands, which are typed into the chat box: given a
    bucket of their own, a client could take turns between `chat_message` and
    `run_command` and talk to the table at twice the rate chat allows.
    """
    event = request.event.get('message') if request else '?'
    args = request.event.get('args') if request else None
    sid = getattr(request, 'sid', '?')
    key = f"{sid}:{budget_event or event}"

    if args and payload_too_large(args[0] if args else None):
        logger.warning("oversized payload for %s from sid=%s", event, sid)
        emit('error', {'code': 'PAYLOAD_TOO_LARGE',
                       'message': 'That message was too large.'})
        return True

    limiters = session()
    budget = limit_for(budget_event or event)
    if limiters.rate_limiter.allow(key, limit=budget):
        return False

    wait = limiters.rate_limiter.retry_after(key, limit=budget)
    strikes = limiters.abuse_tracker.record_violation(sid)
    logger.warning("rate limited %s from sid=%s viewer=%s (strike %s, retry in %.1fs)",
                   event, sid, limiters.viewers.get(sid), strikes, wait)

    if limiters.abuse_tracker.should_disconnect(sid):
        logger.warning("disconnecting sid=%s for sustained flooding", sid)
        emit('error', {'code': 'RATE_LIMITED',
                       'message': 'Too many requests - disconnecting.'})
        disconnect()
        return True

    emit('error', {
        'code': 'RATE_LIMITED',
        'message': f'Slow down - try again in {wait:.1f}s.',
        'details': {'retry_after': wait},
    })
    return True

def log_event(kind, text, player=None, **details):
    """Record something that happened and push it to everyone.

    Timestamps come from the log (server clock) — a client that owns the clock
    could otherwise forge history.
    """
    entry = session().event_log.log(kind, text, player=player, **details)
    socketio.emit('event_logged', {'entry': entry})
    return entry

def viewer_for(sid=None):
    """The player name a socket is currently viewing as, if any."""
    return session().viewers.get(sid or request.sid)

def require_actor(data=None):
    """The seat this socket holds, or None once the action has been refused.

    Every game handler starts here. The acting player is whoever this
    *connection* is seated as; the `name` a client sends is only its claim
    about itself, and until this existed any connected socket could send
    another player's name and act as them. The field is still read — the
    browser puts it in every payload — but only to notice a disagreement,
    which is either a client bug or somebody trying it on.
    """
    seat = session().viewers.get(request.sid)
    claimed = data.get('name') if isinstance(data, dict) else None
    if isinstance(claimed, str) and claimed.strip() and claimed.strip() != seat:
        logger.warning("payload name %r does not match seat %r for %s from sid=%s",
                       claimed, seat, request.event.get('message'), request.sid)

    if seat is None:
        reject('NOT_SEATED', 'Join the table before taking an action')
        return None

    game = session().game
    if game is not None and game.game_state == "started" and not game.is_player(seat):
        # An observer holds a seat in the lobby but none at the table. Without
        # this they would reach the engine under a name it does not know.
        reject('NOT_A_PLAYER', 'Observers cannot act in the game')
        return None
    return seat

def broadcast_board(extra=None):
    """Send the board to every connected socket, filtered per recipient.

    Each socket gets its own payload because hands are hidden information: a
    single broadcast sends identical bytes to everyone, and anything in those
    bytes is readable in DevTools no matter what the UI draws.

    Sent through `socketio.emit` rather than the request-scoped `emit`: the turn
    watchdog broadcasts from a background task with no request behind it, and
    the request-scoped form raised there — so every board update the server sent
    on its own initiative was lost inside the watchdog's except clause.
    """
    live = session()
    if live.game is None:
        return
    for sid, name in list(live.viewers.items()):
        payload = {
            'board': live.game.get_board_data(viewer=name),
            'log_last_id': live.event_log.last_id,
            # Which build served this. A tab left open across a deploy keeps
            # talking to the new server with the old client, and this is what
            # lets it notice: the panel latches the first id it is told and says
            # so, loudly, the moment a later payload disagrees.
            'build': build_info.summary(),
        }
        if extra:
            payload.update(extra)
        socketio.emit('board_updated', payload, to=sid)

def announce_choices():
    """Tell each player who owes a decision what they are being asked.

    Sent to that player's own sockets only, because the options can be the
    contents of somebody else's hand. The table learns that the game is waiting
    from the board snapshot, which carries every pending choice redacted.
    """
    game = session().game
    if game is None or not game.pending_choices:
        return
    for sid, name in list(session().viewers.items()):
        choice = game.pending_choice_for(name)
        if choice is not None:
            # socketio.emit for the same reason as broadcast_board: an
            # expiring choice is announced from the watchdog, with no request.
            socketio.emit('choice_required', game.choice_to_dict(choice, viewer=name), to=sid)

def save_game():
    """Write the game to disk so a restart does not lose it.

    Saved on every accepted change rather than on a timer: the file is small
    (only decided board state, not the derived graph) and this way there is no
    window in which the last move is missing.
    """
    game = session().game
    if game is None:
        return
    try:
        persistence.save(game, SAVE_FILE)
    except Exception:
        # A failed save must never take the live game down with it.
        logger.exception("could not save the game to %s", SAVE_FILE)

def bump_and_broadcast(extra=None):
    """Record that state changed, then push it to everyone."""
    game = session().game
    if game is not None:
        game.state_version += 1
        problems = game.check_invariants()
        if problems:
            # Validation and application disagree — the rules are not actually
            # being enforced where we think they are. Loud, with context.
            logger.error("INVARIANT VIOLATION after %s: %s",
                         request.event.get('message') if request else '?', problems)
        save_game()
    broadcast_board(extra)

def get_user_by_name(users, name):
    for user in users:
        if user.get('name') == name:
            return user
    return None

def remove_user_by_name(users, name):
    return [u for u in users if u.get('name') != name]

def lobby_users():
    """The users actually present right now.

    users.json is a *memory* of everyone who has ever joined — it is what lets
    a returning player keep their colour. It is not the roster. Nothing removes
    an entry when someone closes their tab, so treating the file as the roster
    let four players from a finished session occupy every seat forever while
    the lobby showed nobody. Presence comes from live sockets instead.

    A player who drops mid-game keeps their seat: that roster is the Game
    object, not this list.
    """
    present = set(session().viewers.values())
    return [u for u in load_users() if u.get('name') in present]

def emit_rules(to_sender_only=False):
    """Publish the rule catalogue and the current selection.

    The catalogue travels with the selection so the lobby can render itself
    from the server's registry — adding a rule server-side makes it appear in
    every client with no front-end change. The presets ride along for the same
    reason: each one is a named set of individual rules the lobby can offer as
    a one-click shortcut.
    """
    live = session()
    payload = {
        'catalogue': rules_module.catalogue(),
        'presets': rules_module.presets(),
        'selected': live.lobby_rules,
        'locked': live.game is not None and live.game.game_state == "started",
    }
    if to_sender_only:
        emit('rules_changed', payload)
    else:
        socketio.emit('rules_changed', payload)

    # The command bar renders from the server's catalogue for the same reason
    # the rules picker does, and which commands this table may run is decided by
    # the rules that just went out — so the two travel together rather than
    # leaving the client to work out availability from a rule id.
    emit_commands(to_sender_only)

def emit_commands(to_sender_only=False):
    """Publish the slash command catalogue, availability already worked out."""
    live = session()
    running = live.game is not None and live.game.game_state == "started"
    payload = {
        'commands': commands_module.catalogue(
            live.game.rules if running else live.lobby_rules, running,
        ),
        'prefix': commands_module.PREFIX,
    }
    if to_sender_only:
        emit('commands_changed', payload)
    else:
        socketio.emit('commands_changed', payload)

def send_state_snapshot():
    """Send the full current state to the requesting socket only."""
    game = session().game
    if game is None or game.game_state != "started":
        emit('game_state', {'in_game': False, 'build': build_info.summary()})
        return

    current_player = game.players[game.current_player_index]
    emit('game_state', {
        'in_game': True,
        'build': build_info.summary(),
        'players': game.get_player_names(),
        'observers': game.observers,
        'current_player': current_player.name if current_player else None,
        'board': game.get_board_data(viewer=viewer_for())
    })

def emit_user_list():
    users = lobby_users()
    players = [u for u in users if u.get('role') == 'player']
    observers = [u for u in users if u.get('role') == 'observer']
    socketio.emit('user_list', {
        'players': players,
        'observers': observers,
        'min_players': session().lobby_rules['min_players'],
        'max_players': seat_limit(),
    })

File: server/static/js/state.js
// ------------------------------------------------------------- client state
//
// One object, in two halves.
//
// `server` is the last thing the server said, kept verbatim. Every question the
// server already answers - whose turn it is, whether the dice are up, whether
// the robber or a discard is pending, whether a game is running at all - is
// asked of that half at read time, through the helpers below. Mirroring those
// answers into local flags is what shipped a wedged lobby: a `game_state`
// snapshot that meant "no game is running" was read as "a game started", and
// nothing ever put the flag back.
//
// The rest is genuinely local. Which build button is armed, which knight was
// tapped first, how far the log has been read: the server has no opinion on
// any of it, and no payload can correct it.
export const viewState = {
    server: {
        // The board payload. Null is what "no game is running" means.
        board: null,
        // Who the server seats, and how many it takes to start.
        roster: { players: [], observers: [], minPlayers: 2, maxPlayers: 4 },
        // The lobby rule set, exactly as the server last broadcast it. The
        // client never writes to `selected` itself - a change is emitted and
        // re-read from the `rules_changed` reply, so every table member sees
        // the same thing.
        // `presets` are named sets of individual rules, not modes: picking one
        // ticks its rules and nothing records that it was used.
        rules: { catalogue: [], presets: [], selected: {}, locked: false }
    },

    // What this tab asked to join as. The name is what the player typed; the
    // role and colour are a request, and the roster above is the answer.
    identity: { name: null, requestedRole: null, color: null },

    // 'settlement', 'road', 'city'; with Cities & Knights on - 'knight',
    // 'knight_move' and 'city_wall'; with Seafarers on - 'ship' and
    // 'ship_move'. All of them mean the same thing: the next tap on the board
    // is an intent, not a pan.
    selectedBuilding: null,

    // First half of a knight move. A move is two taps, so the origin has to
    // survive between them.
    knightMoveFrom: null,

    // The same for a ship move: the side the ship was picked up from.
    shipMoveFrom: null,

    // The progress card whose target is being picked on the board, and the
    // targets picked for it so far. Nothing is sent until the last one is in,
    // so a card here is still in the player's hand and cancelling costs them
    // nothing.
    progressPick: { card: null, picked: [] },

    // What the pointer is over, and what a click has offered up for a ✓.
    // Both are `{kind, key, blocked}` or null. `sample` is the raw pointer
    // position, written by pointermove and consumed by the render loop - doing
    // the hit test in the handler would run it far more often than the display
    // can show the result.
    placement: { hover: null, pending: null, sample: null },

    // The gesture in progress on the board, used to tell a tap from a pan.
    pointerDown: null,

    // `game_won` is a one-shot notice that no later payload repeats, so the
    // win has to be latched to stop the turn countdown.
    winnerAnnounced: false,

    // Render loop state - socket handlers set state and mark dirty, never draw.
    render: { dirty: false, highlightNumber: null },

    // `highestId` is what a reconnecting client asks to catch up from, and it
    // is also how a duplicate entry is recognised and dropped.
    log: { highestId: 0 },

    // Rebuilding the rule rows destroys focus and the caret, so they are only
    // rebuilt when the catalogue's shape changes.
    renderedRulesSignature: '',

    // Countdown display only - the server owns expiry. The last `timer` reading
    // it sent, and when it arrived: which clock is running, how much of it is
    // left and how long it was. Null until the first payload says.
    timers: { phase: null, remaining: null, limit: null, updatedAt: Date.now(), handle: null },

    // The once-per-second trade offer countdown.
    tradeTimerHandle: null
};

// ------------------------------------------------ questions the server answers
//
// Each of these reads the stored payload afresh. None of their results may be
// cached in anything that outlives one function call: a cache is a mirror, and
// a mirror is what this section exists to make unnecessary.

/**
 * The board exactly as the server last sent it, or null when no game is on.
 */
export function getBoard() {
    return viewState.server.board;
}

/**
 * What one build costs at this table, or null when the server has not priced it.
 *
 * The prices are the server's, sent with the board and already through every
 * cost modifier the table's rules switched on. The client had its own copy of
 * `server/data/costs.json` until it was deleted: a house rule that changed a
 * price moved the engine's number and not the client's, so the button greyed
 * itself out against a price nobody was charging.
 *
 * @param {string} kind - A build type, as `data/costs.json` names it
 * @returns {object|null} - {resource: amount}
 */
export function getBuildCost(kind) {
    return viewState.server.board?.costs?.[kind] || null;
}

/**
 * Whether there is a game to interact with.
 * A board payload exists for exactly as long as one does - the lobby snapshot
 * and `game_ended` both clear it - and a declared winner ends interaction too.
 */
export function isGameRunning() {
    return viewState.server.board !== null && !viewState.winnerAnnounced;
}

/**
 * 'setup' or 'playing', defaulting to 'playing' the way the payload does.
 */
export function getGamePhase() {
    return viewState.server.board?.game_phase || 'playing';
}

/**
 * Whose turn the server says it is, or null.
 */
export function getCurrentPlayer() {
    return viewState.server.board?.current_player || null;
}

/**
 * Whether the seat this tab is playing is the one on turn.
 */
export function isMyTurn() {
    const me = viewState.identity.name;
    return me !== null && me === getCurrentPlayer();
}

/**
 * Whether the dice are already up this turn.
 */
export function hasRolledDice() {
    return viewState.server.board?.has_rolled_dice === true;
}

/**
 * Whether a 7 has left the robber waiting to be placed.
 */
export function mustMoveRobber() {
    return viewState.server.board?.must_move_robber === true;
}

/**
 * Whether the robber's new hex still needs a victim picked.
 */
export function mustChooseVictim() {
    return viewState.server.board?.must_choose_victim === true;
}

/**
 * Who the robber's hex exposes, as the server worked it out.
 */
export function getRobberVictims() {
    return viewState.server.board?.robber_victims || [];
}

/**
 * How many cards this player owes the bank, or 0.
 */
export function getDiscardAmount() {
    return viewState.server.board?.players_needing_discard?.[viewState.identity.name] || 0;
}

/**
 * The role the server has this player in.
 * Falls back to what the join box asked for until a roster has arrived - a
 * seat is the server's to grant, not the client's to assume.
 */
export function getRole() {
    const roster = viewState.server.roster;
    const me = viewState.identity.name;
    if (roster.players.includes(me)) {
        return 'player';
    }
    if (roster.observers.includes(me)) {
        return 'observer';
    }
    return viewState.identity.requestedRole;
}

/**
 * Record the roster the server just broadcast.
 * The lobby sends user objects and a running game sends bare names, so both
 * shapes are flattened to names here rather than at every read.
 *
 * @param {Array} players - Player names, or user objects carrying `name`
 * @param {Array} observers - The same, for observers
 * @param {number} [minPlayers] - Minimum needed to start, when the payload says
 * @param {number} [maxPlayers] - Seats at this table, when the payload says
 */
export function setRoster(players, observers, minPlayers, maxPlayers) {
    const nameOf = (entry) => (typeof entry === 'string' ? entry : entry?.name);
    viewState.server.roster.players = (players || []).map(nameOf);
    viewState.server.roster.observers = (observers || []).map(nameOf);
    if (typeof minPlayers === 'number') {
        viewState.server.roster.minPlayers = minPlayers;
    }
    // The cap is a rule the table can raise for a 5-6 player board, so the
    // lobby heading reads it from here rather than from a "/4" in the markup.
    if (typeof maxPlayers === 'number') {
        viewState.server.roster.maxPlayers = maxPlayers;
    }
}

/**
 * Fold a partial server message into the stored snapshot.
 *
 * `turn_changed` and `choose_victim` settle facts a moment before the board
 * broadcast that repeats them. Writing them here rather than into a parallel
 * flag keeps one answer to every server-owned question, and the next full board
 * payload replaces the object wholesale.
 *
 * @param {object} facts - Board fields the message settled
 */
export function applyBoardFacts(facts) {
    if (viewState.server.board) {
        Object.assign(viewState.server.board, facts);
    }
}
